endpoint: add optional way of propagating context.Context into regeneration

ianvernon opened the pull request:
All API calls made to the cilium-agent from clients are passed a `context.Context`, which is canceled whenever an API request times out.

When the Cilium CNI plugin is invoked, it performs an API call to the cilium-agent after creating the network device, links, etc. for the pod. This request goes into the `createEndpoint` function. This function initializes an endpoint with the information provided in the API request, and regenerates the endpoint. This includes compiling the program for the endpoint, and interacting with the network device, links, etc. for the pod. If this API request times out, the Cilium CNI plugin will immediately delete the endpoint and all associated network devices, links, etc. for the pod. 

Before this change, the cilium-agent did not stop regeneration if the API request timed out. As a result, the CNI plugin would delete the interfaces for the pod / endpoint while the regeneration was still ongoing. This would cause the datapath / loader interactions to fail for the endpoint, as the devices would no longer exist. This has caused a strain on our CI, as when such interactions with the datapath fail, the datapath packges log an error with the string `JoinEP`, which is searched for in the logs, and fails the CI if found ( https://github.com/cilium/cilium/issues/7513 ). 

To avoid such problems, Cilium needs to be able to know when the context passed in via the API request was canceled. This PR enhances the endpoint regeneration framework to be able to contain a context passed down in the regeneration request types. This context is only populated with the context provided to `createEndpoint` at this time. This context is used as a parent context for the proxy configuration Completion context, as well as the context passed due to the datapath / loader packages. If the context is canceled, we will return the error that the context itself was canceled instead of having `JoinEP` errors in the logs, which will fail the CI. The endpoint regeneration will then fail. This is OK, because it is assumed that the endpoint will be deleted if the context associated with `createEndpoint` is canceled. 

Another nice part of adding context to the endpoint regeneration request, is that this can be passed into the function which acquires the build semaphore for regenerations. If the context is canceled, the semaphore acquisition will be stopped as well. 

Fixes: #7513 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7600)
<!-- Reviewable:end -->

Commit:
endpoint: add Context field to regenerationContext

This will be used to propagate context down from callers of endpoint regenerate
so that endpoint builds can be cancelled in the case of endpoint deletion.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: use parent context with prepareForProxyUpdates

This means that if the context passed down through endpoint regeneration
is cancelled, the proxy configuration will also be cancelled immediately
as well.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: pass down context on endpoint creation into regeneration functionality

Passing down the context from the API request allows for the regeneration to
be cancelled immediately if the associated context is cancelled. This is
safe to do without worrying about an endpoint being in an indeterminate
state in the case of endpoint creation because the endpoint will be deleted
immediately afterwards upon the context being cancelled.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
loader: check whether context is cancelled

This allows for detection of when endpoint builds are
cancelled, instead of trying to perform operations
after operations like endpoint creation have failed.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: pass context down into QueueEndpointBuild

Since regeneration now has context associated with it, we can pass this into
`QueueEndpointBuild`, which releases resources (build semaphore acquisition) if
the endpoint regeneration is cancelled before the actual build permissions have
been acquired. Update unit tests which implement the `Owner` interface to
account for this change as well.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -53,12 +53,22 @@ func (r DatapathRegenerationLevel) String() string {
+	cFunc := e.ParentCancel
+
+	if pCtx == nil {
+		pCtx, cFunc = context.WithCancel(context.Background())
```
Whose responsibility is it to call the cancel function here?

According to the golang documentation in the link below, it appears that there's a goroutine associated with each context and we need to call cancel to prevent goroutine leak.

https://golang.org/pkg/context/#WithCancel

ianvernon replied:
Good find, I never knew this. IMO this should be handled in `endpoint.Regenerate` since the endpoint's EventQueue can be drained, and the actual event handling logic may never occur for an endpoint regeneration that was queued up. It thus has to be the queuer's responsibility.

ianvernon replied:
Should be fixed now :)

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -66,7 +78,12 @@ func reloadDatapath(ctx context.Context, ep endpoint, dirs *directoryInfo) error
 			})
-			scopedLog.WithError(err).Warn("JoinEP: Failed to load program")
+			switch err {
+			case context.Canceled:
```
I'm curious, in `pkg/command/exec/exec.go` we use `context.DeadlineExceeded()` and not `context.Canceled`. I guess that's because typically we've been creating contexts with a timeout and only testing that the timeout condition occurs, whereas now the API could potentially actually invoke the `Cancel()` function first.

I think we need to add this case to `combinedOutput()` in `pkg/command/exec` as well with this PR.

ianvernon replied:
I think I have resolved this!

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -95,6 +101,12 @@ func replaceDatapath(ctx context.Context, ifName string, objPath string, progSec
 func graftDatapath(ctx context.Context, mapPath, objPath, progSec string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
```
If we push this stuff into `pkg/command/exec`, then I'm not sure whether there's a big win from putting these bits at the start of each function -- for example in this case it's only a couple of lines above the `exec.CommandContext()` call where we'd do such a check anyway.

ianvernon replied:
Done! Added additional unit tests in `pkg/command/exec` as well.

joestringer commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -110,9 +127,17 @@ type datapathRegenerationContext struct {
+	if parentCtx != nil {
+		completionCtx, completionCancel = context.WithTimeout(parentCtx, EndpointGenerationTimeout)
+	} else {
+		completionCtx, completionCancel = context.WithTimeout(context.Background(), EndpointGenerationTimeout)
```
When the context is derived from the parent, I'd imagine that the parent's `Cancel()` results in a `Cancel()` of the child. However, in this background case no-one would call the cancel function.

ianvernon replied:
The caller of `prepareForProxyUpdates` calls `defer datapathRegenCtxt.completionCancel()`.

joestringer replied:
Ok yeah. Looks like another potential candidate for doing the context creation in the same layer as the defer to keep management of the context/cancel all in one place.

joestringer reviewed (changes requested):
I'm excited about this, as I think the errors from canceling are going to be more intuitive and make more sense in general than the current approach where the datapath isn't at all aware of cancels happening earlier. That said, I have a few questions below.

Primary concern below is about contexts that are never canceled, which causes goroutine leaks. Incidentally, those docs also suggest to never store contexts in a struct type, as this prevents the static analysis tools from pointing out leaks in code: https://golang.org/pkg/context/

On the datapath side I think we could have very similar behaviour by moving the `ctx.Err() == context.Canceled` handling from each datapath replace function down into `combinedOutput()` in `pkg/command/exec`. We wouldn't need to return a different error like we do in the `context.DeadlineExceeded` case, but we **would** need to ensure that in the case `ctx.Err() == context.Canceled`, we actually return `ctx.Err()`, not the `err` from `cmd.CombinedOutput()`. (That's what the code currently does).

Which brings up the difference between `context.Canceled` and `context.DeadlineExceeded`. Seems like we probably need to handle both in a lot of these scenarios? Do you know whether the API handler stuff provides `WithCancel()` contexts or `WithTimeout()` contexts? (A quick grep seems to suggest it's all doing `WithTimeout()`, so `ctx.Err()` would return `context.DeadlineExceeded` which isn't handled in this PR).

How does testing work for this proposal?

Commit:
exec: return for any error from context

Since contexts passed down to regeneration can now be canceled, we
need to handle not only `DeadlineExceedded` but also `Canceled`.

Signed-off by: Ian Vernon <ian@cilium.dev>

aanm commented on daemon/daemon.go:
```diff
@@ -275,25 +275,26 @@ func (d *Daemon) QueueEndpointBuild(epID uint64) func() {
 	}
 
 	// Acquire succeeded, but the context was canceled after?
 	if ctx.Err() == context.Canceled {
```
should be `ctx.Err != nil`?

aanm commented on pkg/endpoint/events.go:
```diff
@@ -45,7 +45,8 @@ func (ev *EndpointRegenerationEvent) Handle(res chan interface{}) {
 	// being deleted at the same time. More info PR-1777.
-	doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
+	doneFunc, err := owner.QueueEndpointBuild(regenContext.parentContext, uint64(e.ID))
+	// If doneFunc != nil, error is nil.
```
Instead of the comment, please check the error first.

aanm commented on pkg/endpoint/regenerationcontext.go:
```diff
@@ -110,9 +119,17 @@ type datapathRegenerationContext struct {
+		completionCancel context.CancelFunc
+	)
+
+	if parentCtx != nil {
```
When would `parentCtx` be nil?

ianvernon replied:
The behavior has changed now in this PR; It should never be nil at this point. I'll remove the nil check.

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -380,7 +380,7 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
 	defer owner.GetCompilationLock().RUnlock()
 
-	datapathRegenCtxt.prepareForProxyUpdates()
+	datapathRegenCtxt.prepareForProxyUpdates(regenContext.parentContext)
```
Given that you checked `regenMetadata.ParentContext` to be non-nil in an earlier commit I assume this could be nil and thus panic.

It might be easier to make sure that `regenMetadata.ParentContext` is always set and set it to `context.Background()` in `Regenerate()` if not set.

ianvernon replied:
This is what is being done in the code - discussed offline.

tgraf reviewed (approved):
Clarified on slack, code is good to go

raybejjani commented on daemon/daemon.go:
```diff
@@ -243,19 +243,19 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
+		return nil, nil
 	}
 	// Store a cancel function to the 'uniqueID' map so that we can
 	// cancel the wait when the endpoint is being deleted.
```
Does the line below this mess things up? ctx variable is overwritten, isn't it? and then the cancel from the ctx parameter won't ever be seen. In particular, I'm not sure that the `ctx.Err()` check would ever trigger then.

ianvernon replied:
Good catch, fixing.

ianvernon replied:
Fixed.

raybejjani commented on daemon/daemon.go:
```diff
@@ -279,7 +279,7 @@ func (d *Daemon) QueueEndpointBuild(epID uint64) func() {
 
 	// Acquire succeeded, but the context was canceled after?
-	if ctx.Err() == context.Canceled {
+	if ctx.Err() != nil {
```
I might be missing it but, in older code, how would `ctx.Err()` ever be non-nil? It is created with `context.WithCancel(context.Background())`

ianvernon replied:
Fixed this by creating a new context with the `ctx` parameter from `QueueEndpointBuild` as the parent context.

raybejjani commented on pkg/datapath/loader/compile.go:
```diff
@@ -268,7 +268,12 @@ func compileDatapath(ctx context.Context, ep endpoint, dirs *directoryInfo, debu
+				// timed out; this log message should only represent failures
+				// with respect to compiling the program.
+				if ctx.Err() == nil {
+					scopedLog.WithError(err).Debug("JoinEP: Failed to compile")
```
I hate to add more lines of code, but would a debug message when `ctx.Err() != nil` help to track cases where the timeouts happened? I'm not sure if there are other ways to correlate timings or sequences of events.

ianvernon replied:
If the context was canceled while the operation was occurring, the context's error should be propagated via the returned error, which can be logged by callers of the function. I am against this because we will be logging the same error multiple times in the same place, vs. here, the log indicates whether there is a problem with compilation / loading itself failing vs. due to the operation being canceled.
